### PR TITLE
Fixes for `gid_odb_expand_ids`

### PR DESCRIPTION
**Pull request by vmg**
Fixes the issue with leaking the error code for the last OID in the array, and the ODB refreshes that were (still) being performed on missing objects.

### CODE REVIEW

**Comment by vmg** [comment]
cc @ethomson @carlosmn

File: src/odb.c [hunk 1]
```diff
@@ -829,52 +895,32 @@ int git_odb__read_header_or_object(
...
-		raw->data = git__calloc(1, sizeof(uint8_t));
-		return 0;
-	} else if (!git_oid_cmp(id, &empty_tree)) {
-		raw->type = GIT_OBJ_TREE;
-		raw->len = 0;
-		raw->data = git__calloc(1, sizeof(uint8_t));
-		return 0;
-	} else {
-		return GIT_ENOTFOUND;
+	if (error == GIT_PASSTHROUGH) {
```
Comment (vmg) [inline src/odb.c]: Note that we can know distinguish between the backends being unable to read headers, and the backends not finding the header at all.

Also, I've actually implemented ODB refreshing for `git_odb_read_header`. Before that, we wouldn't refresh the ODB on `ENOTFOUND` under no circumstances, we would just fall back to `git_odb_read`, which does refresh.

File: src/odb.c [hunk 2]
```diff
@@ -811,11 +850,38 @@ int git_odb_read_header(size_t *len_p, git_otype *type_p, git_odb *db, const git
...
+
+	for (i = 0; i < db->backends.length && error < 0; ++i) {
+		backend_internal *internal = git_vector_get(&db->backends, i);
+		git_odb_backend *b = internal->backend;
+
+		if (only_refreshed && !b->refresh)
+			continue;
+
+		if (b->read_header != NULL)
+			error = b->read_header(len_p, type_p, b, id);
```
Comment (ethomson) [inline src/odb.c]: We will now stop this loop immediately and return `E_PASSTHROUGH` if any of the `read_header`s returned `E_PASSTHROUGH`.  I feel like the correct thing to do here would be to remember that one of the backends returned `E_PASSTHROUGH`, but continue.  One of the other backends may be the one that actually has this object.

This lets that later-ordered backend service the `read_header` request, on the assumption that `read_header` is cheaper than the full `read`.

**Comment by ethomson** [comment]
@vmg - fixing my bugs since 2003.

Okay, maybe not that long ago, but a more accurate year doesn't rhyme and in any case, it's been a long ass time

**Comment by vmg** [comment]
:joy:

**Pull request merged**